xmtp_mls/src/groups/message_history.rs

```diff
@@ -16,49 +19,79 @@ where
 {
     #[allow(dead_code)]
     pub(crate) async fn send_message_history_request(&self) -> Result<(), GroupError> {
-        let pin_code = "1234".to_string();
-        let request_id = "abc123".to_string();
+        let contents = new_message_history_request();
         let _request = PlaintextEnvelope {
             content: Some(Content::V2(V2 {
-                idempotency_key: String::from("unique"),
-                message_type: Some(HistoryRequest(MessageHistoryRequest {
-                    pin_code,
-                    request_id,
-                })),
+                idempotency_key: new_request_id(),
+                message_type: Some(Request(contents)),
             })),
         };
-        // TODO: Implement sending request to network
+        // TODO: Implement sending this request to the network
         Ok(())
     }
 
     #[allow(dead_code)]
-    pub(crate) async fn send_message_history_response(&self) -> Result<(), GroupError> {
-        let backup_url = "https://example.com/uploads/long-id-123".to_string();
-        let request_id = "abc123".to_string();
-        let backup_file_hash = b"ABC123DEF456";
-        let expiration_time_ns = 123;
+    pub(crate) async fn send_message_history_reply(
+        &self,
+        contents: MessageHistoryReply,
+    ) -> Result<(), GroupError> {
         let _request = PlaintextEnvelope {
             content: Some(Content::V2(V2 {
-                idempotency_key: String::from("unique"),
-                message_type: Some(HistoryResponse(MessageHistoryResponse {
-                    backup_url,
-                    request_id,
-                    backup_file_hash: backup_file_hash.into(),
-                    expiration_time_ns,
-                })),
+                idempotency_key: new_request_id(),
+                message_type: Some(Reply(contents)),
             })),
         };
-        // TODO: Implement sending (responding) request to network
+        // TODO: Implement sending this reply to the network
         Ok(())
     }
 }
 
+pub(crate) fn new_message_history_request() -> MessageHistoryRequest {
+    MessageHistoryRequest {
+        pin_code: new_pin(),
+        request_id: new_request_id(),
+    }
+}
+
+#[allow(dead_code)]
+pub(crate) fn new_message_history_reply(
+    id: &str,
+    url: &str,
+    hash: Vec<u8>,
+    exp: i64,
+) -> MessageHistoryReply {
+    MessageHistoryReply {
+        request_id: id.into(),
+        backup_url: url.into(),
+        backup_file_hash: hash,
+        expiration_time_ns: exp,
+    }
+}
+
+fn new_pin() -> String {
+    let mut rng = rand::thread_rng();
+    let pin: u32 = rng.gen_range(0..10000);
+    format!("{:04}", pin)
+}
+
+fn new_request_id() -> String {
+    Alphanumeric.sample_string(&mut rand::thread_rng(), 24)
```

Comment:
How will this request_id be used?

Reply:
Thinking was that a responder would offer a reply to a specifc msg-history request.
I had thought that using the `idempotency_key` for this could work, but that's serving a different purpose at least for regular messages.

and also its in the original XIP.

We _may_ not need this value in the end; as it may be sufficient that: given any msg-history request that gets sent, and any msg-history reply would have passed through the PIN check to get sent, and  so the requesting installation-id can assume _any_ msg-history reply should be pulled down.